### PR TITLE
[Merged by Bors] - feat: port NumberTheory.LegendreSymbol.AddCharacter

### PR DESCRIPTION
**Pull request by Ruben-VandeVelde**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/NumberTheory/LegendreSymbol/AddCharacter.lean [hunk 1]
```diff
@@ -0,0 +1,465 @@
...
+variable {R : Type u} [AddMonoid R] {R' : Type v} [CommMonoid R']
+
+/-- Interpret an additive character as a monoid homomorphism. -/
+def toMonoidHom : AddChar R R' → Multiplicative R →* R' :=
+  id
+#align add_char.to_monoid_hom AddChar.toMonoidHom
+
+open Multiplicative
+
+-- Porting note: added. Correct?
```
Comment (semorrison) [inline Mathlib/NumberTheory/LegendreSymbol/AddCharacter.lean:102]: Yes!

File: Mathlib/NumberTheory/LegendreSymbol/AddCharacter.lean [hunk 2]
```diff
@@ -0,0 +1,465 @@
...
+
+/-- Define `AddChar R R'` as `(Multiplicative R) →* R'`.
+The definition works for an additive monoid `R` and a monoid `R'`,
+but we will restrict to the case that both are commutative rings below.
+We assume right away that `R'` is commutative, so that `AddChar R R'` carries
+a structure of commutative monoid.
+The trivial additive character (sending everything to `1`) is `(1 : AddChar R R').` -/
+def AddChar : Type max u v :=
+  Multiplicative R →* R'
+-- deriving CommMonoid, Inhabited
```
Comment (semorrison) [inline Mathlib/NumberTheory/LegendreSymbol/AddCharacter.lean:71]:
```suggestion
```

File: Mathlib/NumberTheory/LegendreSymbol/AddCharacter.lean [hunk 3]
```diff
@@ -0,0 +1,465 @@
...
+
+-- porting note: added
+@[simp]
+theorem one_apply (a : R) : (1 : AddChar R R') a = 1 := rfl
+
+instance monoidHomClass : MonoidHomClass (AddChar R R') (Multiplicative R) R' :=
+  MonoidHom.monoidHomClass
+#align add_char.monoid_hom_class AddChar.monoidHomClass
+
+-- Porting note(https://github.com/leanprover-community/mathlib4/issues/5229): added. Correct?
```
Comment (semorrison) [inline Mathlib/NumberTheory/LegendreSymbol/AddCharacter.lean:129]: Yes!

File: Mathlib/NumberTheory/LegendreSymbol/AddCharacter.lean [hunk 4]
```diff
@@ -0,0 +1,465 @@
...
+  cases' hψ with x h
+  use a⁻¹ * x
+  rwa [mulShift_apply, mul_inv_cancel_left₀ ha]
+#align add_char.is_nontrivial.is_primitive AddChar.IsNontrivial.isPrimitive
+
+-- Using `structure` gives a timeout, see
+-- https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/mysterious.20finsupp.20related.20timeout/near/365719262 and
+-- https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/mysterious.20finsupp.20related.20timeout
+-- In Lean4, `set_option genInjectivity false in` may solve this issue.
+-- can't prove that they always exist
```
Comment (semorrison) [inline Mathlib/NumberTheory/LegendreSymbol/AddCharacter.lean:283]: Can you add a porting note here reminding us to revisit this?

File: Mathlib/NumberTheory/LegendreSymbol/AddCharacter.lean (hunk 1 above)
Comment (jcommelin) [reply Mathlib/NumberTheory/LegendreSymbol/AddCharacter.lean:101]:
```suggestion
-- Porting note: added.
```

File: Mathlib/NumberTheory/LegendreSymbol/AddCharacter.lean (hunk 3 above)
Comment (jcommelin) [reply Mathlib/NumberTheory/LegendreSymbol/AddCharacter.lean:128]:
```suggestion
-- Porting note(https://github.com/leanprover-community/mathlib4/issues/5229): added.
```

File: Mathlib/NumberTheory/LegendreSymbol/AddCharacter.lean [hunk 5]
```diff
@@ -0,0 +1,464 @@
...
```
Comment (jcommelin) [inline Mathlib/NumberTheory/LegendreSymbol/AddCharacter.lean:282]:
```suggestion
-- Porting note: Using `structure` gives a timeout, see
-- https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/mysterious.20finsupp.20related.20timeout/near/365719262 and
-- https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/mysterious.20finsupp.20related.20timeout
-- In Lean4, `set_option genInjectivity false in` may solve this issue.
-- can't prove that they always exist
```

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**